### PR TITLE
Sync indexes global option

### PR DESCRIPTION
**Pull request by IslandRhythms**
Why can't I do `await Model.listIndexes().length` off the rip but instead have to save it to a variable? Just curious.

### CODE REVIEW

**Comment by vkarpov15** [comment]
"Why can't I do await Model.listIndexes().length off the rip but instead have to save it to a variable? Just curious."

Order of operations. `Model.listIndexes()` returns a promise, not an array. So `await Model.listIndexes().length` means "get the `.length` property of `Model.listIndexes()`, and `await` on the result", which isn't what you want.

You can do `(await Model.listIndexes()).length`, or `await Model.listIndexes().then(res => res.length)`

File: test/index.test.js [hunk 1]
```diff
@@ -866,5 +866,28 @@ describe('mongoose module:', function() {
...
+      });
+      const syncFalse = db.model('Sync', schema, 'Sync');
+      await syncFalse.init();
+      const Indexes = await syncFalse.listIndexes();
+      console.log(Indexes);
+      assert.equal(Indexes.length, 2);
+      await syncFalse.collection.createIndex({ name: 1 });
+      delete m.connection.models['Sync'];
+      m.set('syncIndexes', true);
+      const syncSchema = new m.Schema({ title: String });
```
Comment (vkarpov15) [inline test/index.test.js:885]: In order to test that `syncIndexes()` correctly adds new indexes as well as drops old indexes, let's add a property that isn't `name` or `title`, and add an index to that new property in the schema.

File: test/index.test.js [hunk 2]
```diff
@@ -866,5 +866,28 @@ describe('mongoose module:', function() {
...
+      const syncTrue = db.model('Sync', syncSchema, 'Sync');
+      await syncTrue.init();
+      const Index = await syncTrue.listIndexes();
+      console.log(Index);
+      assert.equal(Index.length, 1);
```
Comment (vkarpov15) [inline test/index.test.js:890]: Assert that this index is an index on `_id`.

**Review by vkarpov15 — CHANGES_REQUESTED** [review]

**Comment by IslandRhythms** [comment]
Mongoerror, ns not found. The test passes on my machine so I don't know why it's failing here.

File: test/index.test.js [hunk 3]
```diff
@@ -866,5 +866,27 @@ describe('mongoose module:', function() {
...
+      const schema = new m.Schema({
+        title: { type: String, index: true }
+      });
+      const syncFalse = db.model('Sync', schema, 'Sync');
+      await syncFalse.collection.dropIndexes();
+      await syncFalse.init();
+      const Indexes = await syncFalse.listIndexes();
+      assert.equal(Indexes.length, 2);
+      await syncFalse.collection.createIndex({ name: 1 });
+      await m.deleteModel('Sync');
```
Comment (Uzlopak) [inline test/index.test.js:882]: I think this line seems to also force mongoose to drop the collection, which probably does not exist. thats why you get the ns not found error

File: test/index.test.js [hunk 4]
```diff
@@ -866,5 +866,27 @@ describe('mongoose module:', function() {
...
     });
+    it('Allows a syncIndexes global option (gh-11030)', async function() {
+      const m = new mongoose.Mongoose();
+
+      const db = await m.connect('mongodb://localhost:27017/mongoose_test_11030');
+      const schema = new m.Schema({
+        title: { type: String, index: true }
+      });
+      const syncFalse = db.model('Sync', schema, 'Sync');
+      await syncFalse.collection.dropIndexes();
```
Comment (vkarpov15) [inline test/index.test.js:877]: I think you need to get rid of this line to fix the tests. The problem is that the 'Sync' collection doesn't exist yet when you're calling `dropIndexes()` here.

**Review by vkarpov15 — CHANGES_REQUESTED** [review]

**Review by vkarpov15 — APPROVED** [review]
Thanks :+1: will merge into 6.2

**Pull request merged**